Proxy

lpenzey opened the pull request:
This adds the ability to use the server as a proxy server in order to fetch json data from an external api. 

This is done by adding new `routes` with the path defined as the api url minus any parameters, and then pointing those endpoints to a `proxyHandler` that makes the request to the api and returns the body of the response. Then that data is added to the body of the response that the server sends back to the client.

lpenzey committed:
Add api routes with proxy handler

lpenzey committed:
Create api data fetcher

lpenzey committed:
Remove unused imports

lpenzey committed:
Remove params from request path to match routes
If an incoming request has parameters (found by matching against a
question mark and splitting the string) it will grab the base path
(everything before the question mark) in order to find the matching
existing route.

lpenzey committed:
Add proxy response type

rmulhol commented on src/main/java/application/Routes.java:
```diff
@@ -33,7 +34,11 @@
+        router.get("/http://www.ctabustracker.com/bustime/api/v2/getpredictions", Handlers.Proxy);
+        router.get("/http://www.ctabustracker.com/bustime/api/v2/getpatterns", Handlers.Proxy);
+        router.get("/http://www.ctabustracker.com/bustime/api/v2/getstops", Handlers.Proxy);
+        router.get("/http://www.ctabustracker.com/bustime/api/v2/getdirections", Handlers.Proxy);
```
would it be worth masking the proxy a bit? Thinking you could perhaps simplify the url by doing `router.get("/getdirections")` and having your server encapsulate the knowledge of `http://www.ctabustracker.com/bustime/api/v2`.

rmulhol commented on src/test/java/server/RouterTest.java:
```diff
@@ -42,6 +44,22 @@ public void canProcessGETRequestIfRouteHasBeenAdded()  {
+        Request proxyRequest = new Request(new StatusLine(GET, "/http://www.ctabustracker.com/bustime/api/v2/getpredictions", VERSION));
+        Response getResponse = router.generateResponse(proxyRequest);
+
+        assertEquals("HTTP/1.1 200 OK\r\n", getResponse.getStatusLine());
```
What's the object under test for this namespace? Wondering if it's potentially redundant to test a variety of paths if you're stubbing the handler - not that that's a bad thing, but perhaps signals that we could use some other tests to exercise these changes

lpenzey replied:
I meant to update line 57 to have the query parameters in the test request path - I was aiming to test that it would find the match against the defined endpoint in the `router` when the request path has dynamic query parameters.

rmulhol commented on src/main/java/server/response/ResponseTypes.java:
```diff
@@ -50,4 +52,18 @@ public static Response internalError() {
+        return new Response.Builder()
+                .withStatus(VERSION + SP + OK+ CRLF)
+                .withBody(proxyPayload)
+                .build();
```
e.g. what tests that a proxy returns the payload on success and a 500 on failure?

rmulhol commented on src/main/java/server/Router.java:
```diff
@@ -40,30 +41,33 @@ private void addToRoutes(HTTPMethods method, String path, Handler handler) {
     public Response generateResponse(Request request) {
+        String path = request.getRequestPath();
+        if(path.contains("?")){
+            path = path.split("\\?")[0];
```
👍 to detecting query params! Do we do anything with them?

lpenzey replied:
Yes - splitting off the query params was necessary to match the request path with the defined endpoint in the router--so the query params are left in the `request` object that gets passed to `generateResponse` which then uses the query params to make the actual request to the api.

rmulhol commented on src/main/java/server/Router.java:
```diff
@@ -40,30 +41,33 @@ private void addToRoutes(HTTPMethods method, String path, Handler handler) {
 
-        if (!methodAllowed(request)) {
+        if (!methodAllowed(request, path)) {
             HashMap<String, Handler> pathMethods = routes.get(request.getRequestPath());
```
can probably ✂️ some calls to `request.getRequestPath()` now that we've got a `path` var?

rmulhol commented on src/main/java/server/ProxyHandler.java:
```diff
@@ -0,0 +1,30 @@
+        }
+
+        in.close();
+        return content.toString();
```
This is probably hard to test, but also probably something where it's good to think about how we can test it. Feels like a place where a refactor could break things without alarm bells going off

rmulhol reviewed (approved):
This is looking really good! Glad that we have an opportunity to serve our own front end with our own server 👍 

If you've got some time to noodle on this, the main thing I'd think about is how we can augment the test suite to prevent regressions

lpenzey committed:
Add params to proxy test request path

lpenzey committed:
Shorten proxy endpoint path

Formerly the endpoint path for proxy routes was the full api url. Now
the router cleans the requested path before looking up if the path is
defined in routes by splitting on '/' and taking the last element in the
returned collection.
